cpp/ql/src/Security/CWE/CWE-497/SystemData.qll

```diff
@@ -236,15 +236,26 @@ class LogonUser extends SystemData {
   override predicate isSensitive() { any() }
 }
 
-private predicate regQuery(FunctionCall source, VariableAccess use) {
+private newtype TRegQueryParameter =
+  TSubKeyName(Expr e) or
```

Comment:
Could we restrict `e` to be only those arguments that are actually keys? Something like:
```ql
e = any(Call call | call.getTarget.hasName(/* ... */)).getArgument(1)
```
and similarly for the other branches?

Reply:
> That's a lot of unnecessary data!

Do you have any reason to believe the full set of values of the `newtype` would be evaluated?  I would (perhaps naively) expect it not to be, but rather the 'characteristic predicate' (in this case rather trivial) of this type would be added to the various `RegQuery` classes.

I must admit I haven't experimented in this area very much.  It appears you can't create an 'infinite' `newtype` such as:
```
newtype MaybeInt =
  TInt(int i) or
  TNone()
```
which perhaps supports your interpretation.

Reply:
> Do you have any reason to believe the full set of values of the newtype would be evaluated? I would (perhaps naively) expect it not to be, but rather the 'characteristic predicate' (in this case rather trivial) of this type would be added to the various RegQuery classes.

The semantics of `newtype` branches follows the exact same steps as with predicates. `TSubKeyName(Expr e)` is the same as `TSubKeyName(Expr e) { any() }` and contains an `Expr e` for any expression that satisfy the branch's body (which in this case is `any()`). You can see an example of this when you run it:

```cpp
import cpp

newtype TRegQueryParameter =
  TSubKeyName(Expr e) or
  TValueName(Expr e) or
  TReturnData(Expr e)

from int n
where n = strictcount(TRegQueryParameter e)
select n
```
if you look in the log you'll see something like:
```ql
Tuple counts for num#quick-query::TSubKeyName#ff/2@2d8bc2co after 999ms:
  13321297 ~3%     {2} r1 = INVOKE HIGHER-ORDER RELATION construct<quick-query::TRegQueryParameter,0> ON <dom#quick-query::TReturnData#f>
                    return r1
Tuple counts for num#quick-query::TValueName#ff/2@1b66508b after 2s:
  13321297 ~0%     {2} r1 = INVOKE HIGHER-ORDER RELATION construct<quick-query::TRegQueryParameter,1> ON <dom#quick-query::TReturnData#f>
                    return r1
 uple counts for num#quick-query::TReturnData#ff/2@b5e6e6qk after 4s:
  13321297 ~0%     {2} r1 = INVOKE HIGHER-ORDER RELATION construct<quick-query::TRegQueryParameter,2> ON <dom#quick-query::TReturnData#f>
                    return r1
Starting to evaluate predicate #select#query#f#count_range/1@131d37qk
Tuple counts for #select#query#f#count_range/1@131d37qk after 1.2s:
  13321297 ~0%     {1} r1 = SCAN num#quick-query::TSubKeyName#ff OUTPUT In.1 'arg0'
  
  13321297 ~0%     {1} r2 = SCAN num#quick-query::TValueName#ff OUTPUT In.1 'arg0'
  
  13321297 ~2%     {1} r3 = SCAN num#quick-query::TReturnData#ff OUTPUT In.1 'arg0'
  
  26642594 ~0%     {1} r4 = r2 UNION r3
  39963891 ~0%     {1} r5 = r1 UNION r4
                    return r5
```

i.e., each branch will have `13321297` elements.

> I must admit I haven't experimented in this area very much. It appears you can't create an 'infinite' newtype such as:
> ```ql
> newtype MaybeInt =
>   TInt(int i) or
>   TNone()
> ```
> which perhaps supports your interpretation.

Indeed, this would create an infinite branch which is not currently possible.

Reply:
... I've just found and pushed an embarrassingly simple solution.  It isn't _quite_ as clean as before, but its close and still prevents duplication of data in the QL.

Reply:
> ... yep, best I can tell it really is constructed unless we get lucky with the optimizer.

Yeah, I don't know if the optimizer could push in some magic in these branches. I'm not sure I've ever seen that before, though.

> I'll see if I can find a better way to do this.

Great! Do take a look at https://github.com/MathiasVP/ql/commit/b853824aa16b119a3909ed86036695f2710a769f for some inspiration if that helps!